src/sinks/pulsar/config.rs

```diff
@@ -0,0 +1,298 @@
+use crate::sinks::util::TowerRequestConfig;
+use crate::{
+    codecs::EncodingConfig,
+    config::{AcknowledgementsConfig, GenerateConfig, Input, SinkConfig, SinkContext},
+    schema,
+    sinks::{
+        pulsar::sink::{healthcheck, PulsarSink},
+        Healthcheck, VectorSink,
+    },
+};
+use codecs::{encoding::SerializerConfig, TextSerializerConfig};
+use futures_util::FutureExt;
+use pulsar::authentication::oauth2::{OAuth2Authentication, OAuth2Params};
+use pulsar::error::AuthenticationError;
+use pulsar::{
+    compression, message::proto, Authentication, Error as PulsarError, ProducerOptions, Pulsar,
+    TokioExecutor,
+};
+use snafu::ResultExt;
+use value::Kind;
+use vector_common::sensitive_string::SensitiveString;
+use vector_config::configurable_component;
+use vector_core::config::DataType;
+
+/// Configuration for the `pulsar` sink.
+#[configurable_component(sink("pulsar"))]
+#[derive(Clone, Debug)]
+pub struct PulsarSinkConfig {
+    /// The endpoint to which the Pulsar client should connect to.
+    ///
+    /// The endpoint should specify the pulsar protocol and port.
+    #[serde(alias = "address")]
+    #[configurable(metadata(docs::examples = "pulsar://127.0.0.1:6650"))]
+    pub(crate) endpoint: String,
+
+    /// The Pulsar topic name to write events to.
+    #[configurable(metadata(docs::examples = "topic-1234"))]
+    pub(crate) topic: String,
+
+    /// The name of the producer. If not specified, the default name assigned by Pulsar will be used.
+    #[configurable(metadata(docs::examples = "producer-name"))]
+    pub(crate) producer_name: Option<String>,
+
+    /// The log field name or tags key to use for the topic key.
+    ///
+    /// If the field does not exist in the log or in tags, a blank value will be used. If unspecified, the key is not sent.
+    ///
+    /// Pulsar uses a hash of the key to choose the topic-partition or uses round-robin if the record has no key.
+    #[configurable(metadata(docs::examples = "message"))]
+    #[configurable(metadata(docs::examples = "my_field"))]
+    pub(crate) partition_key_field: Option<String>,
+
+    /// The log field name to use for the Pulsar properties.
+    ///
+    /// If omitted, no properties will be written.
+    pub properties_key: Option<String>,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub(crate) batch: BatchConfig,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub compression: PulsarCompression,
+
+    #[configurable(derived)]
+    pub encoding: EncodingConfig,
+
+    #[configurable(derived)]
+    pub(crate) auth: Option<AuthConfig>,
+
+    #[configurable(derived)]
+    #[serde(default)]
+    pub request: TowerRequestConfig,
```

Comment:
It was my understanding that `pulsar` was pretty close to `kafka` with regards to the client, do the normal `request` configuration options make sense?

Reply:
I am not super familiar with either of these sinks, or with defining specific defaults to the tower request settings. It looks like `kafka` sink isn't doing that, from inspecting the code.
I'm not sure how to address this concern. Can you point me to where we handle this for `kafka`, or otherwise how I might ascertain an idea of what values to use ?

Reply:
So my thinking here is it's not really an http service, it's got it's own protocol and client strategies that should be embedded in the `pulsar` crate.

If we look at the [docs](https://pulsar.apache.org/docs/2.11.x/concepts-clients/#client-setup-phase) it does sound similar to the kafka client where it negotiates a TCP target that ends up with a binary protocol (IIRC).

The [`pulsar` crate](https://docs.rs/pulsar/latest/pulsar/producer/struct.ProducerOptions.html) has some options that can be passed in to control batch and compression. It may also handle concurrency/more internally or as part of the protocol itself?

Reply:
Ah, yeah that makes a lot of sense. I suspect the tower request config was added here initially by the original PR author as it is a common setting in new style sinks. I'll yank it.